smallrye-reactive-messaging-kafka-api/src/main/java/io/smallrye/reactive/messaging/kafka/api/KafkaMetadataUtil.java

```diff
@@ -0,0 +1,35 @@
+package io.smallrye.reactive.messaging.kafka.api;
+
+import java.util.Optional;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+/**
+ * Utility to access {@link KafkaMessageMetadata} in a {@link Message}.
+ */
+public class KafkaMetadataUtil {
+
+    /**
+     * Read {@link IncomingKafkaRecordMetadata} from a {@link Message}.
+     *
+     * @param msg the message. Must not be {@code null}
+     * @return The {@link IncomingKafkaRecordMetadata}. May return {@code null}, for example if the message is
+     *         not received from a channel backed by Kafka
+     * @throws NullPointerException if {@code msg} is {@code null}
+     */
+    public static Optional<IncomingKafkaRecordMetadata> readIncomingKafkaMetadata(Message<?> msg) {
+        return msg.getMetadata(IncomingKafkaRecordMetadata.class);
+    }
+
+    /**
+     * Write {@link OutgoingKafkaRecordMetadata} to a {@link Message}.
+     *
+     * @param msg the message
+     * @param outgoingKafkaRecordMetadata the {@link OutgoingKafkaRecordMetadata} to write. Must not be {@code null}.
+     * @throws NullPointerException if {@code msg} or {@code outgoingKafkaRecordMetadata} are {@code null}
+     */
+    public static <T> Message<T> writeOutgoingKafkaMetadata(Message<T> msg,
+            OutgoingKafkaRecordMetadata<T> outgoingKafkaRecordMetadata) {
+        return msg.addMetadata(outgoingKafkaRecordMetadata);
+    }
+}
```

Comment:
It seems this util class is harder to use than the underlying `Message.addMetadata|getMetadata`? :-) Especially on the `write` side -- `Message` is immutable, so adding metadata item returns a new `Message`, which should at the very least be mentioned in a `@return` section in the javadoc.

Reply:
The main reason for this, is that in WildFly/EAP I am not allowed to expose experimental features to users in areas where there is a spec. Essentially we need to stick with the latest version of the spec (i.e. MP RM 2.0). And in areas where there is not a spec, such as the Kafka API, we need to limit the API surface and have something as stable as possible, while avoiding pulling in non-spec things.

Although we use the SmallRye API at runtime, so we have access to e.g.`Metadata`, `Message.add/getMetadata()` etc. when running in the server, that becomes more for internal framework code. We 'hide' these things from users by exposing the spec RM 2.0 API only in our BOMs.

This util class is a way to side-step the problem. It makes it possible to use `Metadata` internally, while not exposing all the other stuff from the SmallRye API module to our users. (Additionally we do some deployment time checks to make sure clever users are not bypassing our restrictions by compiling against the SmallRye API rather then the MP RM 2.0.)

Quarkus, which is less strict in this regard, can continue to use the `Message.add/getMetadata()` methods.

Thanks for pointing out the javadoc, I will make that clearer :)